File: src_python/swap_optimizer.py
"""
To optimize the trees, SCITE-RNA alternates between mutation and cell lineage tree spaces.
"""

import warnings
import numpy as np

from src_python.cell_tree import CellTree
from src_python.mutation_tree import MutationTree
from src_python.utils import load_config_and_set_random_seed

config = load_config_and_set_random_seed()


class SwapOptimizer:
    def __init__(self, sig_digits=9, spaces=None, flipped_mutation_direction=True):
        """
        [Arguments]
            spaces: spaces that will be searched and the order of the search
                    'c' = cell tree space, 'm' = mutation tree space
                    default is ['c','m'], i.e. start with cell tree and search both spaces
            sig_dig: number of significant digits to use when calculating joint probability
        """
        self.mt = None
        self.ct = None
        self.n_decimals = None
        if spaces is None:
            spaces = ["c", "m"]
        self.sig_digits = sig_digits
        self.spaces = spaces
        self.flipped_mutation_direction = flipped_mutation_direction

    @property
    def current_joint(self):
        return round(self.ct.joint, self.n_decimals)

    @property
    def mt_joint(self):
        return round(self.mt.joint, self.n_decimals)

    def postprocess_trees(self):
        """
        Post-process the trees after optimization.
        This includes removing single mutations and updating the joint likelihood.
        """
        self.ct.remove_single_mutations()
        self.mt.remove_single_mutations()
        self.ct.update_all()
        self.mt.update_all()

        # Update the joint likelihoods
        self.ct.update_joint()
        self.mt.update_joint()

    def fit_llh(self, llh_1, llh_2):
        self.ct = CellTree(llh_1.shape[0], llh_1.shape[1], flipped_mutation_direction=self.flipped_mutation_direction)
        self.ct.fit_llh(llh_1, llh_2)

        self.mt = MutationTree(llh_1.shape[1], llh_1.shape[0])
        self.mt.fit_llh(llh_1, llh_2)

        mean_abs = np.sum(np.abs(llh_1 + llh_2)) / 2  # mean abs value when attaching mutations randomly
        self.n_decimals = int(self.sig_digits - np.log10(mean_abs))

    def optimize(self, max_loops=100, reshuffle_nodes=True):
        current_space = 0
        converged = [space not in self.spaces for space in ['c', 'm']]  # choose the spaces to be optimized
        if self.spaces[0] == 'c':  # choose the starting search space
            current_space = 0  # 0 for cell lineage tree, 1 for mutation tree
        elif self.spaces[0] == 'm':
            current_space = 1
        else:
            print("Space has to start with either 'c' or 'm'")

        loop_count = 0
        start_joint = -np.inf

        while not all(converged):
            if loop_count >= max_loops:
                warnings.warn('Maximal loop number exceeded.')
                break
            loop_count += 1

            if current_space == 0:
                print('Optimizing cell lineage tree ...')
                self.ct.exhaustive_optimize()  # loop_count=loop_count)
                self.mt.fit_cell_tree(self.ct)
                self.mt.update_all()

            else:  # i.e. current_space == 1:
                print('Optimizing mutation tree ...')
                self.mt.exhaustive_optimize(prune_single_mutations=reshuffle_nodes)  # loop_count=loop_count)
                self.ct.fit_mutation_tree(self.mt)
                self.ct.update_all()

            print(self.ct.joint, self.mt.joint)

            current_joint = self.current_joint
            if self.spaces == ["m"]:
                current_joint = self.mt_joint
            if start_joint < current_joint:
                converged[current_space] = False
            elif start_joint == current_joint:
                converged[current_space] = True
            else:
                raise RuntimeError('Observed decrease in joint likelihood.')

            start_joint = self.current_joint
            if self.spaces == ["m"]:
                start_joint = self.mt_joint

            if "c" in self.spaces and "m" in self.spaces:  # and converged[current_space] == True
                current_space = 1 - current_space  # switch to the other tree space

        self.ct.postprocess_trees(max_iters=1)

    # def optimize(self, max_loops=100, max_no_improve=3, reshuffle_nodes=True):
    #     """
    #     Joint optimization of cell and mutation trees under a marginalized likelihood model.
    #     Tracks best-scoring parent vectors and restores them at the end.
    #
    #     Parameters:
    #         max_loops (int): Maximum number of iterations.
    #         max_no_improve (int): Stop if no improvement in either tree after these many iterations.
    #         reshuffle_nodes (bool): Whether to allow mutation reshuffling during optimization.
    #     """
    #     current_space = 0 if self.spaces[0] == 'c' else 1
    #     no_improvement = [0, 0]  # [cell tree, mutation tree]
    #     best_joint = [float('-inf'), float('-inf')]  # best likelihood per tree
    #     best_parents = [None, None]  # to store best parent vectors for ct and mt
    #
    #     loop_count = 0
    #     while loop_count < max_loops and (no_improvement[0] < max_no_improve or no_improvement[1] < max_no_improve):
    #         loop_count += 1
    #
    #         if current_space == 0:
    #             print(f"[Loop {loop_count}] Optimizing Cell Lineage Tree ...")
    #             self.ct.exhaustive_optimize()
    #             self.mt.fit_cell_tree(self.ct)
    #             self.mt.update_all()
    #             current_joint = self.mt.joint
    #
    #         else:
    #             print(f"[Loop {loop_count}] Optimizing Mutation Tree ...")
    #             self.mt.exhaustive_optimize(prune_single_mutations=reshuffle_nodes)
    #             self.ct.fit_mutation_tree(self.mt)
    #             self.ct.update_all()
    #             current_joint = self.ct.joint
    #
    #         print(self.ct.joint, self.mt.joint)
    #
    #         if current_joint > best_joint[current_space] + 1e-6:
    #             best_joint[current_space] = current_joint
    #             no_improvement[current_space] = 0
    #             best_parents[current_space] = (
    #                 self.ct.parent_vec if current_space == 0 else self.mt.parent_vec
    #             )
    #         else:
    #             no_improvement[current_space] += 1
    #
    #         # Alternate tree space if both are being optimized
    #         if "c" in self.spaces and "m" in self.spaces:
    #             current_space = 1 - current_space
    #
    #     print("Optimization finished.")
    #     print(f"  Best joint (ct): {best_joint[0]:.6f}")
    #     print(f"  Best joint (mt): {best_joint[1]:.6f}")
    #
    #     # Restore best structures
    #     if best_parents[0] is not None:
    #         self.ct.use_parent_vec(best_parents[0])
    #     if best_parents[1] is not None:
    #         self.mt.use_parent_vec(best_parents[1])
    #
    #     # Final update after restoring structure
    #     self.ct.update_all()
    #     self.mt.update_all()

File: src_cpp/swap_optimizer.cpp
/*
To optimize the trees, SCITE-RNA alternates between mutation and cell lineage tree spaces.
*/

#include <algorithm>

#include "swap_optimizer.h"
#include <utility>
#include "cell_tree.h"

// initialize tree space swap optimizer
SwapOptimizer::SwapOptimizer(std::vector<std::string> spaces, bool reverse_mutations, int n_mut, int n_cells, int sig_digits)
        : sig_digits(sig_digits), spaces(std::move(spaces)), mt(n_mut, n_cells), n_cells_(n_cells), n_mut_(n_mut),
        ct(n_cells, n_mut, reverse_mutations){}

// get rounded cell lineage tree joint LLH
double SwapOptimizer::getCtJoint() const {
    return round_to_n_decimals(ct.joint, n_decimals);
}

// get rounded mutation tree joint LLH
double SwapOptimizer::getMtJoint() const {
    return round_to_n_decimals(mt.joint, n_decimals);
}

// initialize the mutation and cell lineage trees
void SwapOptimizer::fit_llh(const std::vector<std::vector<double>>& llh_1, const std::vector<std::vector<double>>& llh_2) {
    ct.fitLlh(llh_1, llh_2);
    mt.fitLlh(llh_1, llh_2);

    double mean_abs = 0.0;
    for (int i = 0; i < n_cells_; ++i) {
        for (int j = 0; j < n_mut_; ++j) {
            mean_abs += std::abs(llh_1[i][j] + llh_2[i][j]);
        }
    }
    mean_abs /= 2;
    n_decimals = static_cast<int>(sig_digits - std::log10(mean_abs));
}

// optimize mutation and cell lineage tree spaces
void SwapOptimizer::optimize(int max_loops, bool insert_nodes) {
    std::vector<bool> converged = {spaces.end() == std::find(spaces.begin(), spaces.end(), "c"),
                                   spaces.end() == std::find(spaces.begin(), spaces.end(), "m")};

    int current_space = spaces[0] == "c" ? 0 : 1;
    int loop_count = 0;
    double start_joint = -std::numeric_limits<double>::infinity();

    while (!std::all_of(converged.begin(), converged.end(), [](bool v) { return v; })) {
        if (loop_count >= max_loops) {
            std::cerr << "Maximal loop number exceeded." << std::endl;
            break;
        }
        ++loop_count;

        if (current_space == 0) {
            std::cout << "Optimizing cell lineage tree ..." << std::endl;
            ct.exhaustiveOptimize();
            mt.fitCellTree(ct);
            mt.updateAll();
        } else {
            std::cout << "Optimizing mutation tree ..." << std::endl;
            mt.exhaustiveOptimize(insert_nodes);
            ct.fitMutationTree(mt);
            ct.updateAll();
        }

        std::cout << ct.joint << " " << mt.joint << std::endl;

        double current_joint = (spaces.size() == 1 && spaces[0] == "m") ? getMtJoint() :
                               (spaces.size() == 1 && spaces[0] == "c") ? getCtJoint() :
                               (getCtJoint()); // get the likelihood of the current space

        if (start_joint < current_joint) {
            converged[current_space] = false;
        } else if (start_joint == current_joint) {
            converged[current_space] = true;
        } else {
            throw std::runtime_error("Observed decrease in joint likelihood.");
        }

        start_joint = (spaces.size() == 1 && spaces[0] == "m") ? getMtJoint() :
                      (spaces.size() == 1 && spaces[0] == "c") ? getCtJoint() :
                      (getCtJoint()); // get the starting likelihood of the next space


        if (std::find(spaces.begin(), spaces.end(), "c") != spaces.end() &&
            std::find(spaces.begin(), spaces.end(), "m") != spaces.end()) {
            current_space = 1 - current_space;
        }
    }
}

// round to make sure, that smaller rounding errors don't affect the optimization
double SwapOptimizer::round_to_n_decimals(double value, int n_decimals) {
    double scale = std::pow(10.0, n_decimals);
    return std::round(value * scale) / scale;
}